src/main/scala/is/hail/methods/PCRelate.scala

```diff
@@ -29,62 +30,124 @@ object PCRelate {
+
+    val g = vdsToMeanImputedMatrix(vds)
+    val blockedG = BlockMatrix.fromIRM(g, blockSize).cache()
+    val sampleIds = vds.stringSampleIds.toArray[Annotation]
+    val idType = TString()
+
+    val result = new PCRelate(maf, blockSize, statistics, defaultStorageLevel)(hc, blockedG, pcs)
+
+    Table(vds.hc, toRowRdd(result, blockSize, minKinship, statistics), sig, keys)
+      .annotateGlobal(sampleIds.toFastIndexedSeq, TArray(TString()), "sample_ids")
+      .select("{i: global.sample_ids[row.i], j: global.sample_ids[row.j], kin: row.kin, k0: row.k0, k1: row.k1, k2: row.k2}") 
+  }
+
+  // FIXME move matrix formation to Python
+  private def rowsToBDM(x: Array[IndexedSeq[java.lang.Double]]): BDM[Double] = {
+    val nRows = x.length
+    assert(nRows > 0)
+    val nCols = x(0).length
+    val a = new Array[Double](nRows * nCols)
     var i = 0
-    while (i < vds.numCols) {
-      val row = pcs(i).getAs[IndexedSeq[Double]](1)
+    while (i < nRows) {
+      val row = x(i)
+      if (row.length != nCols)
+        fatal(s"pc_relate: column index 0 has $nCols scores but column index $i has ${row.length} scores.")
       var j = 0
-      while (j < k) {
-        scoreArray(j * vds.numCols + i) = row(j)
+      while (j < nCols) {
+        val e = row(j)
+        if (e == null)
+          fatal(s"pc_relate: found missing score at array index $j for column index $i")
+        a(j * nRows + i) = e
         j += 1
       }
       i += 1
     }
-    PCRelate.toTable(vds, new DenseMatrix[Double](vds.numCols, k, scoreArray), maf, blockSize, minKinship, statistics)
+    new BDM(nRows, nCols, a)
   }
 
-  private val signature =
-    TStruct(("i", TString()), ("j", TString()), ("kin", TFloat64()), ("k0", TFloat64()), ("k1", TFloat64()), ("k2", TFloat64()))
-  private val keys = Array("i", "j")
-
-  private def toRowRdd(vds: MatrixTable, pcs: DenseMatrix[Double], maf: Double, blockSize: Int, minKinship: Double, statistics: StatisticSubset): RDD[Row] = {
-    val localSampleIds = vds.stringSampleIds
-    val Result(phi, k0, k1, k2) = new PCRelate(maf, blockSize)(vds, pcs, statistics)
-
-    def fuseBlocks(i: Int, j: Int, lmPhi: DenseMatrix[Double], lmK0: DenseMatrix[Double], lmK1: DenseMatrix[Double], lmK2: DenseMatrix[Double]) = {
-      val iOffset = i * blockSize
-      val jOffset = j * blockSize
-
+  private def toRowRdd(
+    r: Result[M],
+    blockSize: Int,
+    minKinship: Double,
+    statistics: StatisticSubset): RDD[Row] = {
+    
+    def fuseBlocks(i: Int, j: Int,
+      lmPhi: BDM[Double],
+      lmK0: BDM[Double],
+      lmK1: BDM[Double],
+      lmK2: BDM[Double]) = {
+            
       if (i <= j) {
-        val size = lmPhi.rows * lmPhi.cols
-        val ab = new ArrayBuilder[Row]()
-        var jj = 1
+        val iOffset = i * blockSize
+        val jOffset = j * blockSize
+
+        val pairs = new ArrayBuilder[Row]()
+        var jj = 0
         while (jj < lmPhi.cols) {
           var ii = 0
-          // assumes square blocks
```

Comment:
this still assumes square blocks because of the calculation of `rowsAboveDiagonal`, right?